SmartDeviceLink/SDLMenuManager.m

```diff
@@ -0,0 +1,441 @@
+//
+//  SDLMenuManager.m
+//  SmartDeviceLink
+//
+//  Created by Joel Fischer on 4/9/18.
+//  Copyright © 2018 smartdevicelink. All rights reserved.
+//
+
+#import "SDLMenuManager.h"
+
+#import "SDLAddCommand.h"
+#import "SDLAddSubMenu.h"
+#import "SDLArtwork.h"
+#import "SDLConnectionManagerType.h"
+#import "SDLDeleteCommand.h"
+#import "SDLDeleteSubMenu.h"
+#import "SDLError.h"
+#import "SDLFileManager.h"
+#import "SDLImage.h"
+#import "SDLLogMacros.h"
+#import "SDLMenuCell.h"
+#import "SDLMenuParams.h"
+#import "SDLOnCommand.h"
+#import "SDLOnHMIStatus.h"
+#import "SDLRegisterAppInterfaceResponse.h"
+#import "SDLRPCNotificationNotification.h"
+#import "SDLRPCResponseNotification.h"
+#import "SDLSetDisplayLayoutResponse.h"
+#import "SDLVoiceCommand.h"
+
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface SDLMenuCell()
+
+@property (assign, nonatomic) UInt32 parentCellId;
+@property (assign, nonatomic) UInt32 cellId;
+
+@end
+
+@interface SDLVoiceCommand()
+
+@property (assign, nonatomic) UInt32 commandId;
+
+@end
+
+@interface SDLMenuManager()
+
+@property (weak, nonatomic) id<SDLConnectionManagerType> connectionManager;
+@property (weak, nonatomic) SDLFileManager *fileManager;
+
+@property (copy, nonatomic, nullable) SDLHMILevel currentLevel;
+@property (strong, nonatomic, nullable) SDLDisplayCapabilities *displayCapabilities;
+
+@property (strong, nonatomic, nullable) NSArray<SDLRPCRequest *> *inProgressUpdate;
+@property (copy, nonatomic, nullable) SDLMenuUpdateCompletionHandler inProgressHandler;
+@property (assign, nonatomic) BOOL hasQueuedUpdate;
+@property (copy, nonatomic, nullable) SDLMenuUpdateCompletionHandler queuedUpdateHandler;
+@property (assign, nonatomic) BOOL waitingOnHMILevelUpdate;
+
+@property (assign, nonatomic) UInt32 lastMenuId;
+@property (copy, nonatomic) NSArray<SDLMenuCell *> *oldMenuCells;
+@property (assign, nonatomic) BOOL needsUpdate;
+
+@end
+
+UInt32 const ParentIdNotFound = UINT32_MAX;
+
+@implementation SDLMenuManager
+
+- (instancetype)init {
+    self = [super init];
+    if (!self) { return nil; }
+
+    _lastMenuId = 0;
+    _menuCells = @[];
+    _voiceCommands = @[];
+    _oldMenuCells = @[];
+    _needsUpdate = NO;
+
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_registerResponse:) name:SDLDidReceiveRegisterAppInterfaceResponse object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_displayLayoutResponse:) name:SDLDidReceiveSetDisplayLayoutResponse object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_hmiStatusNotification:) name:SDLDidChangeHMIStatusNotification object:nil];
```

Comment:
The manager observes `SDLDidReceiveRegisterAppInterfaceResponse` and `SDLDidReceiveSetDisplayLayoutResponse` but doesn't do anything with information. Is it possible to check the `displayCapabilities` to see if `AddCommands` and `PICS`  support images?